### PR TITLE
Fix `rename_dims` to check for conflicting bin-edge coords

### LINKED ISSUE #2734: rename_dims does not detect clash with bin-edge attr
Given
```.py
ds=sc.Dataset({'a': sc.arange('x', 5), 'b': sc.DataArray(sc.arange('x', 5), attrs={'y': sc.arange('y', 2)})})
```
this does not raise an exception: `ds.rename_dims(x='y')` but actually renames the x dim to y which clashes with the y attr. Copying the result then raises `DimensionError: Cannot add coord 'y' of dims (y: 2) to DataArray with dims (y: 5)`

The rename should detect this clash and raise.

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #2734.

The main change here is to return a new object instead of changing in-place. This matches the Python behavior and made fixing the issue simpler.

### CODE REVIEW

File: docs/about/release-notes.rst [hunk 1]
```diff
@@ -61,6 +61,8 @@ Bugfixes
 * Fixed broken boolean-indexing and sort operations on arrays that contain binned data `#2759 <https://github.com/scipp/scipp/pull/2759>`_.
 * ``transform_coords`` now works with datasets with only coordinates but no data `#2755 <https://github.com/scipp/scipp/pull/2755>`_.
 * Fixed a bug when using :func:`scipp.hist` on a Dataset that contains binned data `#2764 <https://github.com/scipp/scipp/pull/2764>`_.
+* Fixed a in the ``rename`` methods to also rename bin coords and attrs of binned data `#2774 <https://github.com/scipp/scipp/pull/2774>`_.
```
Comment (jl-wynen) [inline docs/about/release-notes.rst:64]:
```suggestion
* Fixed a bug in the ``rename`` methods to also rename bin coords and attrs of binned data `#2774 <https://github.com/scipp/scipp/pull/2774>`_.
```
Same below

File: lib/core/rename.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+#include "scipp-core_export.h"
+#include "scipp/core/except.h"
+#include "scipp/units/dim.h"
+
+namespace scipp::core {
+
+namespace detail {
+template <class T>
+[[nodiscard]] T rename_dims(const T &obj,
+                            const std::vector<std::pair<Dim, Dim>> &names,
```
Comment (jl-wynen) [inline lib/core/rename.h:18]: Use `std::span` instead? (And in functions that use it)

**Review by jl-wynen — COMMENTED** [review]
What impact does this have on performance? Are there any cases that relied on rename to be very cheap because it was in-place? (As opposed to only cheap as it is now)

**Comment by SimonHeybrock** [comment]
> What impact does this have on performance? Are there any cases that relied on rename to be very cheap because it was in-place? (As opposed to only cheap as it is now)

I don't think there are such cases. The main use is likely in relation to coord transform, so processing data is much more expensive.

File: lib/core/rename.h (hunk 2 above)
Comment (SimonHeybrock) [reply lib/core/rename.h:18]: Doesn't that have trouble when passing an initializer_list?

**Comment by SimonHeybrock** [comment]
> > What impact does this have on performance? Are there any cases that relied on rename to be very cheap because it was in-place? (As opposed to only cheap as it is now)
> 
> I don't think there are such cases. The main use is likely in relation to coord transform, so processing data is much more expensive.

On the Python side it was not in-place anyway. So basically the question is whether this is used extensively internally in C++. Since the signature has changed we should see all such uses in the current PR, and I did not notice any.

File: lib/core/rename.h (hunk 2 above)
Comment (jl-wynen) [reply lib/core/rename.h:18]: Yes. Is it used in many places? I only remember seeing it in `groupby`. Not sure if it is worth avoiding the creation of a vector elsewhere. The main use case seems to be in Python where it has to create a vector anyway.

**Review by jl-wynen — APPROVED** [review]

File: lib/core/rename.h (hunk 2 above)
Comment (SimonHeybrock) [reply lib/core/rename.h:18]: Mostly in unit tests.

**Pull request merged**